File: src/pysimlin/simlin/json_types.py
"""Wire-level dataclasses for the simlin patch API.

These types back the JSON patch format understood by the engine
(src/simlin-engine/src/json.rs): patch operations, views, sim specs, and
project structure. Variables themselves are represented by the unified
public types in ``simlin.types`` (Stock, Flow, Aux, Module); the converter
in ``json_converter`` maps those to and from the engine's wire shape.

``Model``/``Project`` here are construction envelopes (``Project.new``
builds its initial empty-project JSON from them) and patch payload carriers
-- NOT a lossless mirror of ``Project.serialize_json()``. In particular the
unified variable types deliberately carry no ``uid`` (the engine preserves
or mints uids on upsert), so structuring a full serialized project through
these types and unstructuring it back would orphan view-element and
loop-metadata uid references. For lossless project persistence, keep the
``serialize_json()`` bytes.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Union

# Runtime import (not TYPE_CHECKING): cattrs resolves these dataclasses'
# string annotations against this module's namespace when structuring the
# patch envelope types.
from .types import Aux, Flow, Module, Stock  # noqa: TC001

# View types


@dataclass
class FlowPoint:
    """A point in a flow's visual representation."""

    x: float
    y: float
    attached_to_uid: int = 0


@dataclass
class LinkPoint:
    """A point in a link's visual representation."""

    x: float
    y: float


@dataclass
class Rect:
    """A rectangle for view bounds."""

    x: float
    y: float
    width: float
    height: float


@dataclass
class StockViewElement:
    """Visual element for a stock."""

    uid: int
    name: str
    x: float
    y: float
    label_side: str = ""


@dataclass
class FlowViewElement:
    """Visual element for a flow."""

    uid: int
    name: str
    x: float
    y: float
    points: list[FlowPoint] = field(default_factory=list)
    label_side: str = ""


@dataclass
class AuxViewElement:
    """Visual element for an auxiliary variable."""

    uid: int
    name: str
    x: float
    y: float
    label_side: str = ""


@dataclass
class CloudViewElement:
    """Visual element for a cloud (source/sink)."""

    uid: int
    flow_uid: int
    x: float
    y: float


@dataclass
class LinkViewElement:
    """Visual element for a causal link."""

    uid: int
    from_uid: int
    to_uid: int
    arc: float | None = None
    multi_points: list[LinkPoint] = field(default_factory=list)


@dataclass
class ModuleViewElement:
    """Visual element for a module."""

    uid: int
    name: str
    x: float
    y: float
    label_side: str = ""


@dataclass
class AliasViewElement:
    """Visual element for an alias (ghost)."""

    uid: int
    alias_of_uid: int
    x: float
    y: float
    label_side: str = ""


# Union type for view elements
ViewElement = Union[
    StockViewElement,
    FlowViewElement,
    AuxViewElement,
    CloudViewElement,
    LinkViewElement,
    ModuleViewElement,
    AliasViewElement,
]


@dataclass
class View:
    """A view/diagram in the model."""

    elements: list[ViewElement] = field(default_factory=list)
    kind: str = ""
    view_box: Rect | None = None
    # Zoom factor (1.0 = 100%, 2.0 = twice as big); 0.0 means the default of 1.0.
    zoom: float = 0.0


# Simulation specs


@dataclass
class SimSpecs:
    """Simulation specification."""

    start_time: float
    end_time: float
    dt: str = ""
    save_step: float = 0.0
    method: str = ""
    time_units: str = ""


# Project structure types


@dataclass
class Dimension:
    """A dimension for subscripted variables."""

    name: str
    elements: list[str] = field(default_factory=list)
    size: int = 0
    maps_to: str | None = None


@dataclass
class Unit:
    """A unit definition."""

    name: str
    equation: str = ""
    disabled: bool = False
    aliases: list[str] = field(default_factory=list)


@dataclass
class LoopMetadata:
    """Metadata for a feedback loop."""

    uids: list[int] = field(default_factory=list)
    deleted: bool = False
    name: str = ""
    description: str = ""


@dataclass
class MacroSpec:
    """Marks a model as a callable macro template and records its calling convention."""

    parameters: list[str] = field(default_factory=list)
    primary_output: str = ""
    additional_outputs: list[str] = field(default_factory=list)


@dataclass
class ModelGroup:
    """Semantic/organizational group for categorizing model variables."""

    name: str
    doc: str | None = None
    parent: str | None = None
    members: list[str] = field(default_factory=list)
    run_enabled: bool = False


@dataclass
class Model:
    """A model in the project."""

    name: str
    stocks: list[Stock] = field(default_factory=list)
    flows: list[Flow] = field(default_factory=list)
    auxiliaries: list[Aux] = field(default_factory=list)
    modules: list[Module] = field(default_factory=list)
    sim_specs: SimSpecs | None = None
    views: list[View] = field(default_factory=list)
    loop_metadata: list[LoopMetadata] = field(default_factory=list)
    groups: list[ModelGroup] = field(default_factory=list)
    macro_spec: MacroSpec | None = None


@dataclass
class Project:
    """A complete system dynamics project."""

    name: str
    sim_specs: SimSpecs
    models: list[Model] = field(default_factory=list)
    dimensions: list[Dimension] = field(default_factory=list)
    units: list[Unit] = field(default_factory=list)


# Patch operation types


@dataclass
class UpsertStock:
    """Operation to insert or update a stock variable."""

    stock: Stock


@dataclass
class UpsertFlow:
    """Operation to insert or update a flow variable."""

    flow: Flow


@dataclass
class UpsertAux:
    """Operation to insert or update an auxiliary variable."""

    aux: Aux


@dataclass
class UpsertModule:
    """Operation to insert or update a module."""

    module: Module


@dataclass
class DeleteVariable:
    """Operation to delete a variable by identifier."""

    ident: str


@dataclass
class RenameVariable:
    """Operation to rename a variable."""

    from_: str  # "from" is a reserved word in Python
    to: str


@dataclass
class UpsertView:
    """Operation to insert or update a view at a specific index."""

    index: int
    view: View


@dataclass
class DeleteView:
    """Operation to delete a view at a specific index."""

    index: int


@dataclass
class UpdateStockFlows:
    """Operation to replace a stock's inflow and outflow lists.

    A targeted alternative to re-upserting the whole stock: only the flow
    wiring changes, every other field of the stock is preserved by the
    engine (``patch.rs::apply_update_stock_flows``).
    """

    ident: str
    inflows: list[str] = field(default_factory=list)
    outflows: list[str] = field(default_factory=list)


@dataclass
class SetLoopName:
    """Operation to pin (name) a feedback loop by its variable set.

    Naming a loop forces the LTM engine to ALWAYS score it -- even in
    discovery mode, whose reported list is capped and may not name it (the
    LOOPSCORE capability from the LTM literature). ``variables`` lists the
    variable names forming the loop's cycle (order does not matter; the cycle
    is recovered from the model's causal graph).
    """

    variables: list[str]
    name: str
    description: str | None = None


@dataclass
class SetSimSpecs:
    """Operation to set simulation specifications."""

    sim_specs: SimSpecs


# Union types for operations
JsonModelOperation = Union[
    UpsertStock,
    UpsertFlow,
    UpsertAux,
    UpsertModule,
    DeleteVariable,
    RenameVariable,
    UpsertView,
    DeleteView,
    UpdateStockFlows,
    SetLoopName,
]

JsonProjectOperation = Union[SetSimSpecs]


# Patch structures


@dataclass
class JsonModelPatch:
    """A patch containing operations for a specific model."""

    name: str
    ops: list[JsonModelOperation] = field(default_factory=list)


@dataclass
class JsonProjectPatch:
    """A patch containing project-level and model-level operations."""

    project_ops: list[JsonProjectOperation] = field(default_factory=list)
    models: list[JsonModelPatch] = field(default_factory=list)

File: src/engine2/src/api/json-types.ts
// Copyright 2025 The Simlin Authors. All rights reserved.
// Use of this source code is governed by the Apache License,
// Version 2.0, that can be found in the LICENSE file.

/**
 * JSON-compatible types for the simlin patch API.
 *
 * These types match the Rust JSON types in src/simlin-engine/src/json.rs
 * and are used for serializing patches to send to the simulation engine.
 */

// Supporting types

/**
 * Scale for graphical function axes.
 */
export interface JsonGraphicalFunctionScale {
  min: number;
  max: number;
}

/**
 * A graphical/table function (lookup table).
 */
export interface JsonGraphicalFunction {
  points?: [number, number][];
  y_points?: number[];
  kind?: string;
  x_scale?: JsonGraphicalFunctionScale;
  y_scale?: JsonGraphicalFunctionScale;
}

/**
 * An element-specific equation for arrayed variables.
 */
export interface JsonElementEquation {
  subscript: string;
  equation: string;
  initial_equation?: string;
  graphical_function?: JsonGraphicalFunction;
}

/**
 * Equation structure for arrayed/subscripted variables.
 */
export interface JsonArrayedEquation {
  dimensions: string[];
  equation?: string;
  initial_equation?: string;
  elements?: JsonElementEquation[];
}

/**
 * A reference mapping between module input/output and parent model variable.
 */
export interface JsonModuleReference {
  src: string;
  dst: string;
}

// Variable types

/**
 * A stock (level, accumulation) variable for JSON serialization.
 */
export interface JsonStock {
  name: string;
  inflows?: string[];
  outflows?: string[];
  uid?: number;
  initial_equation?: string;
  units?: string;
  non_negative?: boolean;
  documentation?: string;
  can_be_module_input?: boolean;
  is_public?: boolean;
  arrayed_equation?: JsonArrayedEquation;
}

/**
 * A flow (rate) variable for JSON serialization.
 */
export interface JsonFlow {
  name: string;
  uid?: number;
  equation?: string;
  units?: string;
  non_negative?: boolean;
  graphical_function?: JsonGraphicalFunction;
  documentation?: string;
  can_be_module_input?: boolean;
  is_public?: boolean;
  arrayed_equation?: JsonArrayedEquation;
}

/**
 * An auxiliary (intermediate calculation) variable for JSON serialization.
 */
export interface JsonAuxiliary {
  name: string;
  uid?: number;
  equation?: string;
  initial_equation?: string;
  units?: string;
  graphical_function?: JsonGraphicalFunction;
  documentation?: string;
  can_be_module_input?: boolean;
  is_public?: boolean;
  arrayed_equation?: JsonArrayedEquation;
}

/**
 * A module (submodel) variable for JSON serialization.
 */
export interface JsonModule {
  name: string;
  model_name: string;
  uid?: number;
  units?: string;
  documentation?: string;
  references?: JsonModuleReference[];
  can_be_module_input?: boolean;
  is_public?: boolean;
}

/**
 * Union type for all JSON variable types.
 */
export type JsonVariable = JsonStock | JsonFlow | JsonAuxiliary | JsonModule;

// View types

/**
 * A point in a flow's visual representation.
 */
export interface JsonFlowPoint {
  x: number;
  y: number;
  attached_to_uid?: number;
}

/**
 * A point in a link's visual representation.
 */
export interface JsonLinkPoint {
  x: number;
  y: number;
}

/**
 * A rectangle for view bounds.
 */
export interface JsonRect {
  x: number;
  y: number;
  width: number;
  height: number;
}

/**
 * Visual element for a stock.
 */
export interface JsonStockViewElement {
  uid: number;
  name: string;
  x: number;
  y: number;
  label_side?: string;
}

/**
 * Visual element for a flow.
 */
export interface JsonFlowViewElement {
  uid: number;
  name: string;
  x: number;
  y: number;
  points?: JsonFlowPoint[];
  label_side?: string;
}

/**
 * Visual element for an auxiliary variable.
 */
export interface JsonAuxiliaryViewElement {
  uid: number;
  name: string;
  x: number;
  y: number;
  label_side?: string;
}

/**
 * Visual element for a cloud (source/sink).
 */
export interface JsonCloudViewElement {
  uid: number;
  flow_uid: number;
  x: number;
  y: number;
}

/**
 * Visual element for a causal link.
 */
export interface JsonLinkViewElement {
  uid: number;
  from_uid: number;
  to_uid: number;
  arc?: number;
  multi_points?: JsonLinkPoint[];
}

/**
 * Visual element for a module.
 */
export interface JsonModuleViewElement {
  uid: number;
  name: string;
  x: number;
  y: number;
  label_side?: string;
}

/**
 * Visual element for an alias (ghost).
 */
export interface JsonAliasViewElement {
  uid: number;
  alias_of_uid: number;
  x: number;
  y: number;
  label_side?: string;
}

/**
 * Union type for view elements.
 */
export type JsonViewElement =
  | JsonStockViewElement
  | JsonFlowViewElement
  | JsonAuxiliaryViewElement
  | JsonCloudViewElement
  | JsonLinkViewElement
  | JsonModuleViewElement
  | JsonAliasViewElement;

/**
 * A view/diagram in the model.
 */
export interface JsonView {
  elements?: JsonViewElement[];
  kind?: string;
  view_box?: JsonRect;
  zoom?: number;
}

// Simulation specs

/**
 * Simulation specification.
 */
export interface JsonSimSpecs {
  start_time: number;
  end_time: number;
  dt?: string;
  save_step?: number;
  method?: string;
  time_units?: string;
}

// Project structure types

/**
 * A dimension for subscripted variables.
 */
export interface JsonDimension {
  name: string;
  elements?: string[];
  size?: number;
  maps_to?: string;
}

/**
 * A unit definition.
 */
export interface JsonUnit {
  name: string;
  equation?: string;
  disabled?: boolean;
  aliases?: string[];
}

/**
 * Metadata for a feedback loop.
 */
export interface JsonLoopMetadata {
  uids?: number[];
  deleted?: boolean;
  name?: string;
  description?: string;
}

/**
 * A model in the project.
 */
export interface JsonModel {
  name: string;
  stocks?: JsonStock[];
  flows?: JsonFlow[];
  auxiliaries?: JsonAuxiliary[];
  modules?: JsonModule[];
  sim_specs?: JsonSimSpecs;
  views?: JsonView[];
  loop_metadata?: JsonLoopMetadata[];
}

/**
 * A complete system dynamics project.
 */
export interface JsonProject {
  name: string;
  sim_specs: JsonSimSpecs;
  models?: JsonModel[];
  dimensions?: JsonDimension[];
  units?: JsonUnit[];
}

// Patch operation payloads

/**
 * Payload for upsert stock operation.
 */
export interface UpsertStockPayload {
  stock: JsonStock;
}

/**
 * Payload for upsert flow operation.
 */
export interface UpsertFlowPayload {
  flow: JsonFlow;
}

/**
 * Payload for upsert auxiliary operation.
 */
export interface UpsertAuxPayload {
  aux: JsonAuxiliary;
}

/**
 * Payload for upsert module operation.
 */
export interface UpsertModulePayload {
  module: JsonModule;
}

/**
 * Payload for delete variable operation.
 */
export interface DeleteVariablePayload {
  ident: string;
}

/**
 * Payload for rename variable operation.
 */
export interface RenameVariablePayload {
  from: string;
  to: string;
}

/**
 * Payload for upsert view operation.
 */
export interface UpsertViewPayload {
  index: number;
  view: JsonView;
}

/**
 * Payload for delete view operation.
 */
export interface DeleteViewPayload {
  index: number;
}

/**
 * Payload for set sim specs operation.
 */
export interface SetSimSpecsPayload {
  sim_specs: JsonSimSpecs;
}

// Operation types with type discriminator

export interface UpsertStockOp {
  type: 'upsert_stock';
  payload: UpsertStockPayload;
}

export interface UpsertFlowOp {
  type: 'upsert_flow';
  payload: UpsertFlowPayload;
}

export interface UpsertAuxOp {
  type: 'upsert_aux';
  payload: UpsertAuxPayload;
}

export interface UpsertModuleOp {
  type: 'upsert_module';
  payload: UpsertModulePayload;
}

export interface DeleteVariableOp {
  type: 'delete_variable';
  payload: DeleteVariablePayload;
}

export interface RenameVariableOp {
  type: 'rename_variable';
  payload: RenameVariablePayload;
}

export interface UpsertViewOp {
  type: 'upsert_view';
  payload: UpsertViewPayload;
}

export interface DeleteViewOp {
  type: 'delete_view';
  payload: DeleteViewPayload;
}

export interface SetSimSpecsOp {
  type: 'set_sim_specs';
  payload: SetSimSpecsPayload;
}

/**
 * Union type for model operations.
 */
export type JsonModelOperation =
  | UpsertStockOp
  | UpsertFlowOp
  | UpsertAuxOp
  | UpsertModuleOp
  | DeleteVariableOp
  | RenameVariableOp
  | UpsertViewOp
  | DeleteViewOp;

/**
 * Union type for project operations.
 */
export type JsonProjectOperation = SetSimSpecsOp;

// Patch structures

/**
 * A patch containing operations for a specific model.
 */
export interface JsonModelPatch {
  name: string;
  ops: JsonModelOperation[];
}

/**
 * A patch containing project-level and model-level operations.
 */
export interface JsonProjectPatch {
  project_ops?: JsonProjectOperation[];
  models?: JsonModelPatch[];
}

// Type guards for discriminating operations

export function isUpsertStock(op: JsonModelOperation): op is UpsertStockOp {
  return op.type === 'upsert_stock';
}

export function isUpsertFlow(op: JsonModelOperation): op is UpsertFlowOp {
  return op.type === 'upsert_flow';
}

export function isUpsertAux(op: JsonModelOperation): op is UpsertAuxOp {
  return op.type === 'upsert_aux';
}

export function isUpsertModule(op: JsonModelOperation): op is UpsertModuleOp {
  return op.type === 'upsert_module';
}

export function isDeleteVariable(op: JsonModelOperation): op is DeleteVariableOp {
  return op.type === 'delete_variable';
}

export function isRenameVariable(op: JsonModelOperation): op is RenameVariableOp {
  return op.type === 'rename_variable';
}

export function isUpsertView(op: JsonModelOperation): op is UpsertViewOp {
  return op.type === 'upsert_view';
}

export function isDeleteView(op: JsonModelOperation): op is DeleteViewOp {
  return op.type === 'delete_view';
}
